src/main/java/at/ac/tuwien/kr/alpha/common/rule/AbstractRule.java

```diff
@@ -0,0 +1,120 @@
+package at.ac.tuwien.kr.alpha.common.rule;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+import at.ac.tuwien.kr.alpha.Util;
+import at.ac.tuwien.kr.alpha.common.atoms.Literal;
+import at.ac.tuwien.kr.alpha.common.rule.head.Head;
+
+/**
+ * An abstract representation of a rule with a specific type of @{link Head} (type parameter H)
+ * 
+ * @param <H> the type of head for this rule Copyright (c) 2019, the Alpha Team.
+ * 
+ *        Copyright (c) 2017-2019, the Alpha Team.
+ */
+public abstract class AbstractRule<H extends Head> {
+
+	private final H head;
+	private final List<Literal> body;
```

Comment:
Thanks for catching this! Storing the body twice is obviously nonsense.. I think I found a neat solution now.

We store positive and negative body `Literal`s as separate `Set<Literal>'s in `AbstractRule`. `AbstractRule#getBody` is realized using Guava's `Sets.union` (no performance overhead becuase the two sets are "stitched together").

`InternalRule`'s methods `List<Atom> getBodyAtomsPositive()` and `List<Atom> getBodyAtomsNegative()` are thereby rendered unnecessary since they`re superseded by `Set<Literal> getPositiveBody()` and `Set<Literal> getNegativeBody()` in `AbstractRule`.

Storing positive and negative bodies separately from the beginning eliminates any need to save the same literal more than once. Also, I think working on `Set`s rather than `List`s here makes more sense since rule bodies are only ever iterated over and should be duplicate-free (any concerns with regards to iteration order are, I think, dismissible since - in the cases where it matters - iteration order is determined by a `RuleGroundingOrder` anyway).
In addition, by using `Literal`s as the stored type rather than `Atom`s, I was able to eliminate a few lines like
```
for(Atom a : ruleBody){
    Literal lit = a.toLiteral();
    ...
}
```